crates/syn2mas/src/synapse_reader.rs

```diff
@@ -0,0 +1,281 @@
+}
+
+#[derive(Iden)]
+pub enum ExtraSynapseUserIden {
+    AppserviceId,
+    IsGuest,
+}
+
+/// List of Synapse tables that we should acquire an `EXCLUSIVE` lock on.
+///
+/// This is a safety measure against other processes changing the data underneath our feet.
+/// It's still not a good idea to run Synapse at the same time as the migration.
+// TODO not complete!
+const TABLES_TO_LOCK: &[&str] = &["users"];
+
+/// Number of migratable rows in various Synapse tables.
+/// Used to estimate progress.
+#[derive(Clone, Debug)]
+pub struct SynapseRowCounts {
+    pub users: i64,
+}
+
+pub struct SynapseReader<'c> {
+    conn: &'c mut PgConnection,
+}
+
+impl<'conn> SynapseReader<'conn> {
+    /// Create a new Synapse reader, which entails creating a transaction and locking Synapse tables.
+    ///
+    /// # Errors
+    ///
+    /// Errors are returned under the following circumstances:
+    ///
+    /// - An underlying database error
+    /// - If we can't lock the Synapse tables (pointing to the fact that Synapse may still be running)
+    pub async fn new(
+        synapse_connection: &'conn mut PgConnection,
+        dry_run: bool,
+    ) -> Result<Self, Error> {
+        query("BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE READ ONLY DEFERRABLE;")
+            .execute(&mut *synapse_connection)
+            .await
+            .into_database("begin transaction")?;
+
+        let lock_type = if dry_run {
+            // We expect dry runs to be done alongside Synapse running, so we don't want to
+            // interfere with Synapse's database access in that case.
+            "ACCESS SHARE"
+        } else {
+            "EXCLUSIVE"
+        };
+        for table in TABLES_TO_LOCK {
+            query(&format!("LOCK TABLE {table} IN {lock_type} MODE NOWAIT;"))
+                .execute(&mut *synapse_connection)
+                .await
+                .into_database_with(|| format!("locking Synapse table `{table}`"))?;
+        }
+
+        Ok(Self {
+            conn: synapse_connection,
+        })
+    }
+
+    /// Finishes the Synapse reader, committing the transaction.
+    ///
+    /// # Errors
+    ///
+    /// Errors are returned under the following circumstances:
+    ///
+    /// - An underlying database error whilst committing the transaction.
+    pub async fn finish(self) -> Result<(), Error> {
+        // TODO enforce that this is called somehow.
+
+        query("COMMIT;")
+            .execute(self.conn)
+            .await
+            .into_database("end transaction")?;
+        Ok(())
+    }
+
+    /// Counts the rows in the Synapse database to get an estimate of how large the migration is going to be.
+    ///
+    /// # Errors
+    ///
+    /// Errors are returned under the following circumstances:
+    ///
+    /// - An underlying database error
+    pub async fn count_rows(&mut self) -> Result<SynapseRowCounts, Error> {
+        // TODO no need for query builder here
```

Comment:
I assume that if we wanted to use sqlx here, we'd need to split synapse reading logic to its own crate, and somehow setup a synapse-like schema for introspection?

Reply:
*Probably* more effort than it's worth, it gets quite fiddly to maintain the different schemas :(